### PR TITLE
TTL support (and a few enhancements)

### PR DESCRIPTION
**Pull request by nathan-cormier**
Introduce an extensible Options struct as a param for NewStream, which includes a TTL param. TTL leverages Redis Streams MINID param to expire old stream entries.

Also:
* a constant for NoExpiration when TTL isn't desired
* add RevRange method to Stream
* support time.Time and time.Duration for ser/de
* new Metadata field type for generic structured data
* updated README

### CODE REVIEW

**Comment by nathan-cormier** [comment]
Hi @dranikpg - I have a few enhancements to your package that I've been using from my fork, but I thought I'd see if you're interested in merging them back into your original package. Let me know what you think of these, thanks!

File: utils.go [hunk 1]
```diff
@@ -2,19 +2,26 @@ package gtrs
...
 	"strings"
+	"time"
 
 	"github.com/redis/go-redis/v9"
 )
 
+// Metadata is a type that allows serialization of generic structured
+// metadata within the stream entries. Any value that can be serialized
+// to JSON can be inserted here.
+type Metadata map[string]any
```
Comment (dranikpg) [inline utils.go:20]: Cool idea. Maybe add a line about it in the readme?

File: stream.go [hunk 2]
```diff
@@ -2,19 +2,40 @@ package gtrs
...
 
 // Create a new stream with messages of type T.
-func NewStream[T any](client redis.Cmdable, stream string) Stream[T] {
-	return Stream[T]{client: client, stream: stream}
+// Options are optional (the parameter can be nil to use defaults).
+func NewStream[T any](client redis.Cmdable, stream string, opt *Options) Stream[T] {
+	ttl := NoExpiration
+	if opt != nil {
+		ttl = opt.TTL
+	}
```
Comment (dranikpg) [inline stream.go:37]: Not against the nil option. Libraries for example (go redis) usually expect a non nil pointer for all cases... But lets maybe keep it 🤷🏻 🙂

File: stream.go [hunk 3]
```diff
@@ -63,6 +94,27 @@ func (s Stream[T]) Range(ctx context.Context, from, to string, count ...int64) (
...
+	if err != nil {
+		return nil, ReadError{Err: err}
+	}
+
+	msgs := make([]Message[T], len(redisSlice))
+	for i, msg := range redisSlice {
+		msgs[i] = toMessage[T](msg, s.stream)
+	}
+	return msgs, nil
+}
```
Comment (dranikpg) [inline stream.go:116]: I thought of re-using one internal function that takes client methods for it, but its likely not worth it

**Review by dranikpg — APPROVED** [review]
Wow! Awesome contribution 🔥 Thanks 🙇🏻‍♀️

Your implementation is so clean I have nothing to comment about 😆 My only suggestion would be adding a mention of the new Metadata type to the readme

Otherwise I'll release soon 0.3 (or 1.0 if the interface changed 🤔)... Anyways a new version with your changes 😅

File: utils.go (hunk 1 above)
Comment (nathan-cormier) [reply utils.go:20]: Sounds good, added a README section for this 👍

File: stream.go (hunk 2 above)
Comment (nathan-cormier) [reply stream.go:37]: Yeah trying to be defensive here and also make it slightly easier for callers who have no options to specify. I don't feel too strongly about it either, but it's already in there and used by the unit tests, so if you're OK with it, I am too 😉

**Pull request merged**